### PR TITLE
[#1277] feat(API): Add column attribute autoIncrement of the increment interface.

### PR DESCRIPTION
**Pull request by Clearvive**
### What changes were proposed in this pull request?
Add column attribute autoIncrement of the increment interface.

### Why are the changes needed?
Fix: #1277

### Does this PR introduce _any_ user-facing change?
  1. Change in user-facing APIs.

### How was this patch tested?
UT

### CODE REVIEW

**Review by mchades — COMMENTED** [review]
Should you add an exception thrown in the server side implementation since the catalog has not been implemented yet?

**Comment by mchades** [comment]
> > Should you add an exception thrown in the server side implementation since the catalog has not been implemented yet?
> 
> It is not easy to determine whether it is set by the user in the catalog layer, as it is not a wrapper Boolean object

we can only throw an exception when `autoIncrement()==true`, because `autoIncrement()==false` is naturally supported.

File: api/src/test/java/com/datastrato/gravitino/rel/TestTransforms.java [hunk 1]
```diff
@@ -50,6 +50,14 @@ public String comment() {
...
             return true;
           }
+
+          @Override
+          public boolean autoIncrement() {
+            return false;
+          }
+
+          @Override
+          public void validate() throws IllegalArgumentException {}
```
Comment (mchades) [inline api/src/test/java/com/datastrato/gravitino/rel/TestTransforms.java:60]: It's not necessary to override this method.

File: api/src/test/java/com/datastrato/gravitino/rel/TestTransforms.java [hunk 2]
```diff
@@ -97,6 +105,14 @@ public String comment() {
...
```
Comment (mchades) [inline api/src/test/java/com/datastrato/gravitino/rel/TestTransforms.java:115]: ditto

File: common/src/main/java/com/datastrato/gravitino/dto/requests/TableCreateRequest.java [hunk 3]
```diff
@@ -108,5 +110,13 @@ public void validate() throws IllegalArgumentException {
...
       Arrays.stream(partitioning).forEach(p -> p.validate(columns));
     }
+
+    List<ColumnDTO> autoIncrementCols =
+        Arrays.stream(columns)
+            .peek(ColumnDTO::validate)
+            .filter(ColumnDTO::autoIncrement)
+            .collect(Collectors.toList());
+    Preconditions.checkArgument(
+        autoIncrementCols.size() <= 1, "Only one column can be auto-incremented.");
```
Comment (mchades) [inline common/src/main/java/com/datastrato/gravitino/dto/requests/TableCreateRequest.java:120]: Listing the column names of auto-incremented columns in the error message would be more user-friendly.

File: api/src/main/java/com/datastrato/gravitino/rel/types/Types.java [hunk 4]
```diff
@@ -923,4 +923,12 @@ public int hashCode() {
       return Arrays.hashCode(types);
     }
   }
+
+  /**
+   * @param dataType The data type to check.
+   * @return True if the given data type is allowed to be an auto-increment column.
+   */
+  public static boolean allowAutoIncrement(Type dataType) {
+    return dataType instanceof IntegerType || dataType instanceof LongType;
```
Comment (yuqi1129) [inline api/src/main/java/com/datastrato/gravitino/rel/types/Types.java:932]: Other int types like `smallint` and `tinyint` must also be considered. 

image

File: integration-test/build.gradle.kts [hunk 5]
```diff
@@ -283,7 +283,7 @@ tasks.test {
 
       // Gravitino CI Docker image
       environment("GRAVITINO_CI_HIVE_DOCKER_IMAGE", "datastrato/gravitino-ci-hive:0.1.7")
-      environment("GRAVITINO_CI_TRINO_DOCKER_IMAGE", "datastrato/gravitino-ci-trino:0.1.2")
+      environment("GRAVITINO_CI_TRINO_DOCKER_IMAGE", "datastrato/gravitino-ci-trino:0.1.3")
```
Comment (qqqttt123) [inline integration-test/build.gradle.kts:286]: Do you release a Docker image? If you release a Docker image, you should add the document in the https://github.com/datastrato/gravitino/blob/main/docs/docker-image-details.md

**Comment by jerryshao** [comment]
@mchades can you please take care of this PR? Thanks.

**Comment by mchades** [comment]
Overall, LGTM, but some comments left and IT issue seem to be addressed.

File: api/src/main/java/com/datastrato/gravitino/rel/Column.java [hunk 6]
```diff
@@ -30,5 +30,21 @@ public interface Column {
   /** @return True if this column may produce null values. Default is true. */
   boolean nullable();
 
+  /** @return True if this column is an auto-increment column. Default is false. */
+  boolean autoIncrement();
+
+  default void validate() throws IllegalArgumentException {
```
Comment (jerryshao) [inline api/src/main/java/com/datastrato/gravitino/rel/Column.java:36]: What is the purpose of adding this interface here?

File: integration-test/build.gradle.kts (hunk 5 above)
Comment (jerryshao) [reply integration-test/build.gradle.kts:286]: Please do it. BTW, why do you need to upgrade trino ci version?

**Comment by qqqttt123** [comment]
If this PR change user-facing interface, we should also add the related documents.

File: api/src/main/java/com/datastrato/gravitino/rel/Column.java (hunk 6 above)
Comment (jerryshao) [reply api/src/main/java/com/datastrato/gravitino/rel/Column.java:36]: Is it necessary to add to the API, maybe adding to ColumnDTO is enough?

**Review by jerryshao — APPROVED** [review]

**Pull request merged**